jams/sonify.py

```diff
@@ -0,0 +1,113 @@
+#!/usr/bin/env python
+# CREATED:2015-12-12 18:20:37 by Brian McFee <brian.mcfee@nyu.edu>
+r'''
+Sonification
+============
+
+.. autosummary::
+    :toctree: generated/
+
+    sonify
+'''
+
+import six
+import numpy as np
+import mir_eval.sonify
+from mir_eval.util import filter_kwargs
+from .eval import coerce_annotation
+from .exceptions import NamespaceError
+
+__all__ = ['sonify']
+
+
+
+def clicks(annotation, sr=22050, length=None, **kwargs):
+    '''Sonify clicks timings
+
+    '''
+
+    interval, _ = annotation.data.to_interval_values()
+
+    return filter_kwargs(mir_eval.sonify.clicks, interval[:, 0],
+                         fs=sr, length=length, **kwargs)
+
+
+def chord(annotation, sr=22050, length=None, **kwargs):
+    '''Sonify chords'''
+
+    intervals, chords = annotation.data.to_interval_values()
+
+    return filter_kwargs(mir_eval.sonify.chords,
+                         chords, intervals,
+                         fs=sr, length=length,
+                         **kwargs)
+
+
+def pitch_hz(annotation, sr=22050, length=None, **kwargs):
+    '''Sonify pitches in Hz'''
+
+    intervals, pitches = annotation.data.to_interval_values()
+
+    # Collapse down to a unique set of frequency values
+    freqs = np.unique(pitches)
+
+    # Build the piano roll
+    pitch_index = {p: i for i, p in enumerate(freqs)}
+    gram = np.zeros((len(freqs), len(pitches)))
+    for t, n in enumerate(pitches):
+        gram[pitch_index[n], t] = 1.0
+
+    return filter_kwargs(mir_eval.sonify.time_frequency,
+                         gram, freqs, intervals,
+                         fs=sr, length=length,
+                         **kwargs)
+
+
+SONIFY_MAPPING = {'beat': clicks,
+                  'segment_open': clicks,
+                  'onset': clicks,
+                  'chord': chord,
+                  'pitch_hz': pitch_hz}
+
+
+def sonify(annotation, sr=22050, duration=None, **kwargs):
+    '''Sonify a jams annotation through mir_eval
+
+    Parameters
+    ----------
+    annotation : jams.Annotation
+        The annotation to sonify
+
+    sr = : int > 0
+        The sampling rate of the output waveform
+
+    duration : float (optional)
+        Optional length (in seconds) of the output waveform
+
+    kwargs
+        Additional keyword arguments to mir_eval.sonify functions
+
+    Returns
+    -------
+    y_sonified : np.ndarray
+        The waveform of the sonified annotation
+
+    Raises
+    ------
+    NamespaceError
+        If the annotation has an un-sonifiable namespace
+    '''
+
+    length = None
+    if duration is not None:
+        length = int(duration * sr)
+
+    for namespace, func in six.iteritems(SONIFY_MAPPING):
+        try:
+            coerce_annotation(annotation, namespace)
+            return func(annotation, sr=sr, length=length, **kwargs)
+        except NamespaceError:
+            pass
```

Comment:
> Proposed solution: use an ordered dict to store the mappings, and put the most specific namespaces first. The conversion logic stays the same, but sonification is always done with the most specific applicable method. If you want to sonify using an alternate method, you can always auto-convert the annotation before sonifying.

This could work, but the logic could get increasingly unwieldy as we add sonifications/namespaces/conversions. For example, in the future we might want to add destructive one-way conversions such as going from pitch_midi to onsets. The sonifications are dramatically different, and since onsets might come first in the ordered dictionary, it could get messy?

I actually don't think that a direct dictionary mapping from each namespace-to-sonification would be that bad... we don't have that many namespaces and it's not a quantity that's likely to grow very fast. And anyone adding a new namespace can make a PR that just adds the new namespace to the mapping dictionary, and since every namespace is mapped independently we don't have to worry about some new namespace or new conversion breaking the logic of the ordered dict. I actually think it's the simplest solution in the long run, but well, open to rebuttals as always.